### ISSUE TITLE
pythonLay can not work

### ISSUE THREAD

**Issue by Cv9527** [BVLC/caffe#5567]
Please use the [caffe-users list](https://groups.google.com/forum/#!forum/caffe-users) for usage, installation, or modeling questions, or other requests for help.
_Do not post such requests to Issues._ Doing so interferes with the development of Caffe.

Please read the [guidelines for contributing](https://github.com/BVLC/caffe/blob/master/CONTRIBUTING.md) before submitting this issue.

### Issue summary
I0426 17:23:21.631978 11182 caffe.cpp:218] Using GPUs 0
I0426 17:23:21.656056 11182 caffe.cpp:223] GPU 0: TITAN X (Pascal)
I0426 17:23:22.040179 11182 solver.cpp:44] Initializing solver from parameters: 
base_lr: 0.0001
display: 1000
max_iter: 500000
lr_policy: "step"
gamma: 0.8
momentum: 0.9
weight_decay: 0.004
stepsize: 10000
snapshot: 10000
snapshot_prefix: "./models/"
solver_mode: GPU
device_id: 0
net: "train12.prototxt"
train_state {
  level: 0
  stage: ""
}
I0426 17:23:22.041339 11182 solver.cpp:87] Creating training net from net file: train12.prototxt
I0426 17:23:22.057746 11182 net.cpp:51] Initializing net from parameters: 
name: "face_12c"
state {
  phase: TRAIN
  level: 0
  stage: ""
}
layer {
  name: "PythonLayer"
  type: "Python"
  top: "data"
  top: "label"
  top: "roi"
  python_param {
    module: "pythonLayer"
    layer: "Data_Layer_train"
  }
}
layer {
  name: "conv1"
  type: "Convolution"
  bottom: "data"
  top: "conv1"
  param {
    lr_mult: 1
  }
  param {
    lr_mult: 2
  }
  convolution_param {
    num_output: 10
    kernel_size: 3
    weight_filler {
      type: "gaussian"
      std: 0.0001
    }
    bias_filler {
      type: "constant"
    }
  }
}
layer {
  name: "relu1"
  type: "PReLU"
  bottom: "conv1"
  top: "conv1"
}
layer {
  name: "pool1"
  type: "Pooling"
  bottom: "conv1"
  top: "pool1"
  pooling_param {
    pool: MAX
    kernel_size: 3
    stride: 2
  }
}
layer {
  name: "conv2"
  type: "Convolution"
  bottom: "pool1"
  top: "conv2"
  param {
    lr_mult: 1
  }
  param {
    lr_mult: 2
  }
  convolution_param {
    num_output: 16
    kernel_size: 3
    weight_filler {
      type: "gaussian"
      std: 0.0001
    }
    bias_filler {
      type: "constant"
    }
  }
}
layer {
  name: "relu2"
  type: "PReLU"
  bottom: "conv2"
  top: "conv2"
}
layer {
  name: "conv3"
  type: "Convolution"
  bottom: "conv2"
  top: "conv3"
  param {
    lr_mult: 1
    decay_mult: 1
  }
  param {
    lr_mult: 2
    decay_mult: 0
  }
  convolution_param {
    num_output: 32
    kernel_size: 3
    stride: 1
    weight_filler {
      type: "xavier"
    }
    bias_filler {
      type: "constant"
      value: 0
    }
  }
}
layer {
  name: "relu3"
  type: "PReLU"
  bottom: "conv3"
  top: "conv3"
}
layer {
  name: "conv4-1"
  type: "Convolution"
  bottom: "conv3"
  top: "conv4-1"
  param {
    lr_mult: 1
    decay_mult: 1
  }
  param {
    lr_mult: 2
    decay_mult: 0
  }
  convolution_param {
    num_output: 2
    kernel_size: 1
    stride: 1
    weight_filler {
      type: "xavier"
    }
    bias_filler {
      type: "constant"
      value: 0
    }
  }
}
layer {
  name: "cls_bridge"
  type: "Python"
  bottom: "conv4-1"
  bottom: "label"
  top: "conv4-1-valid"
  top: "label-valid"
  python_param {
    module: "pythonLayer"
    layer: "cls_Layer_fc"
  }
}
layer {
  name: "cls_loss"
  type: "SoftmaxWithLoss"
  bottom: "conv4-1-valid"
  bottom: "label-valid"
  top: "cls_loss"
  loss_weight: 1
  propagate_down: true
  propagate_down: false
}
layer {
  name: "cls_Acc"
  type: "Accuracy"
  bottom: "conv4-1-valid"
  bottom: "label-valid"
  top: "cls_Acc"
  include {
    phase: TRAIN
  }
}
layer {
  name: "conv4-2"
  type: "Convolution"
  bottom: "conv3"
  top: "conv4-2"
  param {
    lr_mult: 1
    decay_mult: 1
  }
  param {
    lr_mult: 2
    decay_mult: 0
  }
  convolution_param {
    num_output: 4
    kernel_size: 1
    stride: 1
    weight_filler {
      type: "xavier"
    }
    bias_filler {
      type: "constant"
      value: 0
    }
  }
}
layer {
  name: "roi_loss"
  type: "Python"
  bottom: "conv4-2"
  bottom: "roi"
  top: "roi_loss"
  loss_weight: 0.5
  python_param {
    module: "pythonLayer"
    layer: "regression_Layer"
  }
}
I0426 17:23:22.057878 11182 layer_factory.hpp:77] Creating layer PythonLayer
*** Aborted at 1493198605 (unix time) try "date -d @1493198605" if you are using GNU date ***
PC: @     0x7f5cdd4da873 std::_Hashtable<>::clear()
*** SIGSEGV (@0x9) received by PID 11182 (TID 0x7f5d0faab840) from PID 9; stack trace: ***
    @     0x7f5cfff01390 (unknown)
    @     0x7f5cdd4da873 std::_Hashtable<>::clear()
    @     0x7f5cdd4cc346 google::protobuf::DescriptorPool::FindFileByName()
    @     0x7f5cdd4aaac8 google::protobuf::python::cdescriptor_pool::AddSerializedFile()
    @     0x7f5d00eac7d0 (unknown)
    @     0x7f5d00fd501c (unknown)
    @     0x7f5d00f2b3dd (unknown)
    @     0x7f5d00efe1e3 (unknown)
    @     0x7f5d00f1eae5 (unknown)
    @     0x7f5d00eb5123 (unknown)
    @     0x7f5d00efe1e3 (unknown)
    @     0x7f5d00ea913c (unknown)
    @     0x7f5d00fd501c (unknown)
    @     0x7f5d00ea3b89 (unknown)
    @     0x7f5d00f381b4 (unknown)
    @     0x7f5d00f38b8f (unknown)
    @     0x7f5d00f3a300 (unknown)
    @     0x7f5d00f3a5c8 (unknown)
    @     0x7f5d00f3b6db (unknown)
    @     0x7f5d00eb2698 (unknown)
    @     0x7f5d00efe1e3 (unknown)
    @     0x7f5d00fd4447 (unknown)
    @     0x7f5d00ea75c6 (unknown)
    @     0x7f5d00fd501c (unknown)
    @     0x7f5d00ea3b89 (unknown)
    @     0x7f5d00f381b4 (unknown)
    @     0x7f5d00f38b8f (unknown)
    @     0x7f5d00f3a300 (unknown)
    @     0x7f5d00f3a5c8 (unknown)
    @     0x7f5d00f3b6db (unknown)
    @     0x7f5d00eb2698 (unknown)
    @     0x7f5d00efe1e3 (unknown)
Segmentation fault (core dumped)

### Steps to reproduce
I define pythonLayer for my training task, when launch training, it encounter the errors. Is any one can help me to solve it?  

If you are having difficulty building Caffe or training a model, please ask the caffe-users mailing list. If you are reporting a build error that seems to be due to a bug in Caffe, please attach your build configuration (either Makefile.config or CMakeCache.txt) and the output of the make (or cmake) command.

### Your system configuration
Operating system:ubuntu 16.4
Compiler:i7
CUDA version (if applicable):cuda8.0
CUDNN version (if applicable):
BLAS:
Python or MATLAB version (for pycaffe and matcaffe respectively):

**Comment by williford** [comment]
Ask on StackOverflow and/or the Caffe Users mailing list. You can post the link of your question in this thread. If you then email me at jonathan@neural.vision, I can make sure to help you at one of these external sites.

You may also find @shaibagon's tutorial helpful:
http://stackoverflow.com/questions/41344168/what-is-a-python-layer-in-caffe/41481539#41481539

**Comment by Cv9527** [comment]
Thank you @williford　Ｉ have read @shaibagon's tutoria and check my pythonLayer, but the problem is still not solved. Additional, does pythonLayer supports multi-GPU, since my machine has three Titan x.
my pythonLayer define as follow:
import sys
sys.path.append('/home/cmcc/caffe-master/python')
import cv2
import caffe
import numpy as np
import random
import cPickle as pickle
imdb_exit = True

def view_bar(num, total):
    rate = float(num) / total
    rate_num = int(rate * 100)
    r = '\r[%s%s]%d%%' % ("#"*rate_num, " "*(100-rate_num), rate_num, )
    sys.stdout.write(r)
    sys.stdout.flush()
################################################################################
#########################Data Layer By Python###################################
################################################################################
class Data_Layer_train(caffe.Layer):
    def setup(self, bottom, top):
        self.batch_size = 64
	net_side = 12
	cls_list = ''
	roi_list = ''
	pts_list = ''
	cls_root = ''
	roi_root = ''
	pts_root = ''
        self.batch_loader = BatchLoader(cls_list,roi_list,pts_list,net_side,cls_root,roi_root,pts_root)
        top[0].reshape(self.batch_size, 3, net_side, net_side)
        top[1].reshape(self.batch_size, 1)
	top[2].reshape(self.batch_size, 4)
	top[3].reshape(self.batch_size, 10)
    def reshape(self, bottom, top):
        pass

    def forward(self, bottom, top):
	loss_task = random.randint(0,2)
        for itt in range(self.batch_size):
            im, label, roi, pts= self.batch_loader.load_next_image(loss_task)
            top[0].data[itt, ...] = im
            top[1].data[itt, ...] = label
	    top[2].data[itt, ...] = roi
	    top[3].data[itt, ...] = pts
    def backward(self, top, propagate_down, bottom):
        pass

class BatchLoader(object):
    def __init__(self,cls_list,roi_list,pts_list,net_side,cls_root,roi_root,pts_root):
	self.mean = 128
        self.im_shape = net_side
        self.cls_root = cls_root
	self.roi_root = roi_root
	self.pts_root = pts_root
	self.roi_list = []
	self.cls_list = []
	self.pts_list = []
	print "Start Reading Classify Data into Memory..."
	if imdb_exit:
	    fid = open('12/cls.imdb','r')
	    self.cls_list = pickle.load(fid)
	    fid.close()
	else:
	    fid = open(cls_list,'r')
            lines = fid.readlines()
	    fid.close()
	    cur_=0
	    sum_=len(lines)
	    for line in lines:
	        view_bar(cur_, sum_)
	        cur_+=1
	        words = line.split()
	        image_file_name = self.cls_root + words[0] + '.jpg'
	        im = cv2.imread(image_file_name)
	        h,w,ch = im.shape
	        if h!=self.im_shape or w!=self.im_shape:
	            im = cv2.resize(im,(int(self.im_shape),int(self.im_shape)))
	        im = np.swapaxes(im, 0, 2)
	        im -= self.mean
                label    = int(words[1])
                roi      = [-1,-1,-1,-1]
		pts      = [-1,-1,-1,-1,-1,-1,-1,-1,-1,-1]
	        self.cls_list.append([im,label,roi,pts])
	random.shuffle(self.cls_list)
        self.cls_cur = 0
	print "\n",str(len(self.cls_list))," Classify Data have been read into Memory..."

	print "Start Reading Regression Data into Memory..."
	if imdb_exit:
	    fid = open('12/roi.imdb','r')
	    self.roi_list = pickle.load(fid)
	    fid.close()
	else:
	    fid = open(roi_list,'r')
            lines = fid.readlines()
	    fid.close()
	    cur_=0
	    sum_=len(lines)
	    for line in lines:
	        view_bar(cur_, sum_)
	        cur_+=1
	        words = line.split()
	        image_file_name = self.roi_root + words[0] + '.jpg'
	        im = cv2.imread(image_file_name)
	        h,w,ch = im.shape
	        if h!=self.im_shape or w!=self.im_shape:
	            im = cv2.resize(im,(int(self.im_shape),int(self.im_shape)))
	        im = np.swapaxes(im, 0, 2)
	        im -= self.mean
                label    = int(words[1])
                roi      = [float(words[2]),float(words[3]),float(words[4]),float(words[5])]
		pts      = [-1,-1,-1,-1,-1,-1,-1,-1,-1,-1]
	        self.roi_list.append([im,label,roi,pts])
	random.shuffle(self.roi_list)
	self.roi_cur = 0 
	print "\n",str(len(self.roi_list))," Regression Data have been read into Memory..."

	print "Start Reading pts-regression Data into Memory..."
	if imdb_exit:
	    fid = open('12/pts.imdb','r')
	    self.pts_list = pickle.load(fid)
	    fid.close()
	else:
	    fid = open(pts_list,'r')
            lines = fid.readlines()
	    fid.close()
	    cur_=0
	    sum_=len(lines)
	    for line in lines:
	        view_bar(cur_, sum_)
	        cur_+=1
	        words = line.split()
	        image_file_name = self.pts_root + words[0] + '.jpg'
	        im = cv2.imread(image_file_name)
	        h,w,ch = im.shape
	        if h!=self.im_shape or w!=self.im_shape:
	            im = cv2.resize(im,(int(self.im_shape),int(self.im_shape)))
	        im = np.swapaxes(im, 0, 2)
	        im -= self.mean
                label    = -1
                roi      = [-1,-1,-1,-1]
		pts	 = [float(words[ 6]),float(words[ 7]),
			    float(words[ 8]),float(words[ 9]),
			    float(words[10]),float(words[11]),
			    float(words[12]),float(words[13]),
			    float(words[14]),float(words[15])]
	        self.pts_list.append([im,label,roi,pts])
	random.shuffle(self.pts_list)
	self.pts_cur = 0 
	print "\n",str(len(self.pts_list))," pts-regression Data have been read into Memory..."

    def load_next_image(self,loss_task): 
	if loss_task == 0:
	    if self.cls_cur == len(self.cls_list):
                self.cls_cur = 0
                random.shuffle(self.cls_list)
            cur_data = self.cls_list[self.cls_cur]  # Get the image index
	    im       = cur_data[0]
            label    = cur_data[1]
            roi      = [-1,-1,-1,-1]
	    pts	     = [-1,-1,-1,-1,-1,-1,-1,-1,-1,-1]
	    if random.choice([0,1])==1:
		im = cv2.flip(im,random.choice([-1,0,1]))
            self.cls_cur += 1
            return im, label, roi, pts

	if loss_task == 1:
	    if self.roi_cur == len(self.roi_list):
                self.roi_cur = 0
                random.shuffle(self.roi_list)
	    cur_data = self.roi_list[self.roi_cur]  # Get the image index
	    im	     = cur_data[0]
            label    = -1
            roi      = cur_data[2]
	    pts	     = [-1,-1,-1,-1,-1,-1,-1,-1,-1,-1]
            self.roi_cur += 1
            return im, label, roi, pts

	if loss_task == 2:
	    if self.pts_cur == len(self.pts_list):
                self.pts_cur = 0
                random.shuffle(self.pts_list)
	    cur_data = self.pts_list[self.pts_cur]  # Get the image index
	    im	     = cur_data[0]
            label    = -1
            roi      = [-1,-1,-1,-1]
	    pts	     = cur_data[3]
            self.pts_cur += 1
            return im, label, roi, pts
################################################################################
#########################ROI Loss Layer By Python###############################
################################################################################
class regression_Layer(caffe.Layer):
    def setup(self,bottom,top):
	if len(bottom) != 2:
	    raise Exception("Need 2 Inputs")
    def reshape(self,bottom,top):
	if bottom[0].count != bottom[1].count:
	    raise Exception("Input predict and groundTruth should have same dimension")
	roi = bottom[1].data
	self.valid_index = np.where(roi[:,0] != -1)[0]
	self.N = len(self.valid_index)
        self.diff = np.zeros_like(bottom[0].data, dtype=np.float32)
        top[0].reshape(1)

    def forward(self,bottom,top):
	self.diff[...] = 0
	top[0].data[...] = 0
	if self.N != 0:
	    self.diff[...] = bottom[0].data - np.array(bottom[1].data).reshape(bottom[0].data.shape)
            top[0].data[...] = np.sum(self.diff**2) / bottom[0].num / 2.

    def backward(self,top,propagate_down,bottom):
	for i in range(2):
	    if not propagate_down[i] or self.N==0:
		continue
	    if i == 0:
		sign = 1
	    else:
		sign = -1
	    bottom[i].diff[...] = sign * self.diff / bottom[i].num
################################################################################
#############################SendData Layer By Python###########################
################################################################################
class cls_Layer_fc(caffe.Layer):
    def setup(self,bottom,top):
	if len(bottom) != 2:
	    raise Exception("Need 2 Inputs")
    def reshape(self,bottom,top):
	label = bottom[1].data
	self.valid_index = np.where(label != -1)[0]
	self.count = len(self.valid_index)
	top[0].reshape(len(bottom[1].data), 2,1,1)
	top[1].reshape(len(bottom[1].data), 1)
    def forward(self,bottom,top):
	top[0].data[...][...]=0
	top[1].data[...][...]=0
	top[0].data[0:self.count] = bottom[0].data[self.valid_index]
	top[1].data[0:self.count] = bottom[1].data[self.valid_index]
    def backward(self,top,propagate_down,bottom):
	if propagate_down[0] and self.count!=0:
	    bottom[0].diff[...]=0
	    bottom[0].diff[self.valid_index]=top[0].diff[...]
	if propagate_down[1] and self.count!=0:
	    bottom[1].diff[...]=0
	    bottom[1].diff[self.valid_index]=top[1].diff[...]

**Comment by shaibagon** [comment]
Please use the [caffe-users list][1] for usage, installation, or modeling questions, or other requests for help.
You may also post questions on [stackoverflow][3], make sure you tag them with `caffe` tag.  
Do not post such requests to Issues. Doing so interferes with the development of Caffe.

Please read the [guidelines for contributing][2] before submitting this issue.

 [1]: https://groups.google.com/forum/#!forum/caffe-users
 [2]: https://github.com/BVLC/caffe/blob/master/CONTRIBUTING.md
 [3]: http://stackoverflow.com/questions/tagged/caffe

**Comment by linsonwang** [comment]
@Cv9527 hi, cv9527, did you solve this issue?
